core/market/src/db/dao/agreement.rs

```diff
@@ -113,10 +117,52 @@ impl<'c> AgreementDao<'c> {
         .await
     }
 
-    pub async fn approve(&self, id: &AgreementId) -> Result<(), StateError> {
+    pub async fn confirm(
+        &self,
+        id: &AgreementId,
+        session: &AppSessionId,
+    ) -> Result<(), StateError> {
         let id = id.clone();
+        let session = session.clone();
+
         do_with_transaction(self.pool, move |conn| {
-            let agreement: Agreement = dsl::market_agreement.filter(dsl::id.eq(&id)).first(conn)?;
+            let agreement: Agreement =
+                market_agreement.filter(agreement::id.eq(&id)).first(conn)?;
+
+            if agreement.state != AgreementState::Proposal {
+                Err(StateError::InvalidTransition {
+                    id: id.clone(),
+                    from: agreement.state,
+                    to: AgreementState::Pending,
+                })?
+            }
+
+            update_state(conn, &id, &AgreementState::Pending)
+                .map_err(|e| StateError::DbError(e.into()))?;
+
+            if let Some(session) = session {
+                diesel::update(market_agreement.filter(agreement::id.eq(&id)))
```

Comment:
In `update_state` you are using shorter form. Is there a difference?
```suggestion
                diesel::update(market_agreement.find(id))
```

Reply:
I'm not sure, if there's a difference. I didn't wrote this part of `update_state`.

Reply:
Hmm I checked. `find` will return single record and `filter` can return multiple records. So this will be correct here, I will change.